[UnifiedPDF] Only update platform highlight data for page(s) of interest

aprotyas opened the pull request:
#### 94e2e0c433d20ef173dd08f404c1d3603b29c0db
<pre>
[UnifiedPDF] Only update platform highlight data for page(s) of interest
<a href="https://bugs.webkit.org/show_bug.cgi?id=271203">https://bugs.webkit.org/show_bug.cgi?id=271203</a>
<a href="https://rdar.apple.com/124976470">rdar://124976470</a>

Reviewed by Simon Fraser.

DDHighlightData is required for hit testing and to reflect zooming or
scrolling. As such, we only care about updating platform highlight data
on the page where we need to handle a mouse event, or on the page where
a zoom or scroll operation occurred.

This patch updates didInvalidateHighlightOverlayRects to accept an
optional page index to limit which page&apos;s platform highlight data
requires updating. If the caller fails to provide an index, or if that
index does not already exist in our highlight map, we fallback to the
existing behavior of updating the highlight data on all pages.

This is a performance optimization for large PDF files since we avoid
unnecessarily iterating through all pages containing data detector
results.

* Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/PDFDataDetectorOverlayController.h:
* Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/PDFDataDetectorOverlayController.mm:
(WebKit::PDFDataDetectorOverlayController::handleMouseEvent):
(WebKit::PDFDataDetectorOverlayController::updatePlatformHighlightData):
(WebKit::PDFDataDetectorOverlayController::didInvalidateHighlightOverlayRects):
* Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/UnifiedPDFPlugin.mm:
(WebKit::UnifiedPDFPlugin::didChangeScrollOffset):

Canonical link: <a href="https://commits.webkit.org/276329@main">https://commits.webkit.org/276329@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/af0aca25a379604dce2e73d4f2cba3a78579b725

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/44350 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/23418 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/46785 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/47003 "Built successfully") | [  ~~🛠 wincairo~~](https://ews-build.webkit.org/#/builders/32/builds/40379 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/27398 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/20814 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/36511 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/44926 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/20481 "Passed tests") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/18/builds/38186 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-wpe~~](https://ews-build.webkit.org/#/builders/41/builds/17552 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/42/builds/17932 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | [✅ 🛠 wpe-skia](https://ews-build.webkit.org/#/builders/52/builds/2402 "Built successfully") | 
| [⏳ 🛠 🧪 jsc ](https://ews-build.webkit.org/#/builders/JSC-Tests-EWS "Waiting in queue, processing has not started yet") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/40532 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/39578 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/48618 "Built successfully") | 
| [⏳ 🛠 🧪 jsc-arm64 ](https://ews-build.webkit.org/#/builders/JSC-Tests-arm64-EWS "Waiting in queue, processing has not started yet") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/19330 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/50/builds/15882 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/43408 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🛠 tv-sim~~](https://ews-build.webkit.org/#/builders/46/builds/20690 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/21/builds/42136 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/9865 "Built successfully and passed tests") | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/43/builds/21018 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/20318 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

rr-codes commented on Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/PDFDataDetectorOverlayController.mm:
```diff
@@ -237,15 +238,24 @@
+            return m_pdfDataDetectorItemsWithHighlightsMap.keys();
+        }();
+
+        WTF::forEach(pageIndices, [this](auto pageIndex) {
```
why can't this be a normal `for x : pageIndices` loop?

aprotyas replied:
No real reason, I'd just written it that way. I can simplify it to `for x : pageIndices`

aprotyas replied:
(Will do so in an upcoming patch...)

rr-codes commented on Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/PDFDataDetectorOverlayController.mm:
```diff
@@ -237,15 +238,24 @@
     if (shouldUpdatePlatformHighlightData == ShouldUpdatePlatformHighlightData::Yes) {
-        WTF::forEach(m_pdfDataDetectorItemsWithHighlightsMap.keys(), [&](auto pageIndex) {
+        auto pageIndices = [&] {
+            if (pageIndex) {
```
maybe this should be
```
if (!pageIndex) 
  return  m_pdfDataDetectorItemsWithHighlightsMap.keys();

// etc
```

to avoid extra nesting?

aprotyas replied:
I went with this first, but then settled for what exists currently because otherwise you need three `return` statements, and the tail looks awkward here.

```cpp
if (!pageIndex)
    return map.keys();

if (map.find(*pageIndex))
    return customRange(...pageIndex...);

return map.keys();
```

aprotyas committed:
[UnifiedPDF] Only update platform highlight data for page(s) of interest
https://bugs.webkit.org/show_bug.cgi?id=271203
rdar://124976470

Reviewed by Simon Fraser.

DDHighlightData is required for hit testing and to reflect zooming or
scrolling. As such, we only care about updating platform highlight data
on the page where we need to handle a mouse event, or on the page where
a zoom or scroll operation occurred.

This patch updates didInvalidateHighlightOverlayRects to accept an
optional page index to limit which page's platform highlight data
requires updating. If the caller fails to provide an index, or if that
index does not already exist in our highlight map, we fallback to the
existing behavior of updating the highlight data on all pages.

This is a performance optimization for large PDF files since we avoid
unnecessarily iterating through all pages containing data detector
results.

* Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/PDFDataDetectorOverlayController.h:
* Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/PDFDataDetectorOverlayController.mm:
(WebKit::PDFDataDetectorOverlayController::handleMouseEvent):
(WebKit::PDFDataDetectorOverlayController::updatePlatformHighlightData):
(WebKit::PDFDataDetectorOverlayController::didInvalidateHighlightOverlayRects):
* Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/UnifiedPDFPlugin.mm:
(WebKit::UnifiedPDFPlugin::didChangeScrollOffset):

Canonical link: https://commits.webkit.org/276329@main
